### ISSUE TITLE
Leader commits log entries of older terms

### ISSUE THREAD

**Issue by meteam2022** [bakwc/PySyncObj#169]
## Describe the Bug

According to the raft author [Ongaro's dissertation](https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf) (Section 3.6.2 Committing entries from previous terms), leader cannot determine commitment using log entries from older terms.

![Screenshot_20221127_121417](https://user-images.githubusercontent.com/118866424/204118905-c4eba5c8-1e60-460a-b8f5-986d945d5c8e.png)

However, the commitment code does not check the log entry's term. 

https://github.com/bakwc/PySyncObj/blob/8ea2fd6d5f01f2ba413a3ec821883b3883a328cb/pysyncobj/syncobj.py#L576-L588

## To Reproduce

Assume there is a 3-node cluster (N1, N2, and N3), and the network has large delay.

`msg[1-9]+` represents messages between N1-N3.
`msg[a-z]+` represents messages between N2-N3.
`client-op` appends a log entry to the leader's log.

1. **N3 becomes Leader (term 1)**, and sends `msg1 ae` to N1 and `msga ae` to N2.
2. N2 receives `msga ae` and replies `msgb nni success`.
3. N3 makes a `client-op`, and sends `msg2 ae` and `msgc ae`.
4. N3 receives `msgb nni success`.
5. N2 receives `msgc ae` and replies `msgd nni success`.
6. N3 receives `msgd nni success`
7. N1 receives `msg1 ae` and `msg2 ae`. And N1 replies `msg3 nni success` and `msg4 nni success`.
8. N3 makes a `client-op`, and sends `msg5 ae` and **`msge ae entries=[1..4]`**.
9. N3 receives `msg3 nni success` and `msg4 nni success`.
10. **N1 election time out (term 2)**, and sends `msg6 rv` to N3 (and `msg rv` to N2).
11. N1 receives `msg5 ae`, and does not reply because msg's term is smaller.
12. N3 receives `msg6 rv`, but does not vote for N1 because the log of N1 is older.
13. **N3 election time out (term 3)**, and sends `msg7 rv` to N1 (and `msg rv` to N2).
14. N1 receives `msg7 rv`, and replies `msg8 granted` to N3.
15. N3 receives `msg8 granted` and **becomes leader**.
16. N2 receives **`msge ae entries=[1..4]`** and replies `msgf nni=5 success` to N3. (Assume #167 is fixed and line 911 is changed to `nextNodeIdx = newEntries[-1][1] + 1`.  If we don't apply this change, N2 replies `msgf nni=4 success`, which won't affect the result).
17. N3 receives `msgf nni=5 success`.
18. N3 advances commit index to 4. The log entry's term of idx 4 is 1, but N3's term is 3, which violates the specification.

## Possible Fixes

1. Simply add a term check of line 584 and 585. 
1. Iterate from the end of the log to uncommitted log entries in current term.
2. Move the commitment code to `message['type'] == 'next_node_idx'` handling. If the log entry's term equals to current term, check match index and commit if majority nodes match. It should be validated since it changes the core logic.

**Comment by bakwc** [comment]
According to raft paper:

      If there exists an N such that N > commitIndex, a majority
      of matchIndex[i] ≥ N, and log[N].term == currentTerm:
      set commitIndex = N (§5.3, §5.4)

So it's not enough to add a term check. I should also rewrite a while loop and check all indices from currently committed to the last one and try to find at least one possible. And not break at line 587 - right?

**Comment by tangruize** [comment]
Raft ensures the `Log Matching` safety:
> Log Matching: if two logs contain an entry with the same
> index and term, then the logs are identical in all entries up
> through the given index. §5.3

If the last majority matching log entry has the same term with current term, it can be committed, and all log entries before the log entry are also committed.

So I think it is enough to add a term check:

```python
         if count > (len(self.__otherNodes) + 1) / 2: 
             IF log entry term == current term:  # pseudo-code 
                 self.__raftCommitIndex = nextCommitIndex 
                 self.__raftLog.setRaftCommitIndex(self.__raftCommitIndex) 
                 self.__leaderCommitIndex = self.__raftCommitIndex 
         else: 
             break
```

After the check, the loop does not break.

Because of `Log Matching` safety property, we can optimize the loop: iterate backwards, find first commit-able log entry and break. And it should be ok to place checking commitment code on receivig `next_node_idx` msg. It avoids loops.

**Comment by bakwc** [comment]
@meteam2022 please review the fix, https://github.com/bakwc/PySyncObj/pull/170

**Comment by bakwc** [comment]
Fixed by #170
